### PR TITLE
Support ARM 64-bit architecture

### LINKED ISSUE #221: ARM support
It would be nice to provide an efficient execution on ARM. This architecture is widespread on mobile devices and will be used for future Apple Mac CPUs. AWS also provides [instances based on ARM](https://aws.amazon.com/ec2/graviton/).

To do:

1. Figure out what is required to cross-compile to ARM.
1. Look into the [ARM Compute Library](https://github.com/ARM-software/ComputeLibrary) which has GEMM primitives optimized for ARM NEON.
1. Add ARM NEON vectorization for CPU kernels, and update automatic ISA dispatch accordingly.

### PR DESCRIPTION
**Pull request by keichi**
Closes #221

- Add OpenBLAS and Apple Accelerate GEMM backends (DNNL backend also works on ARM)
- Add vectorized kernels using NEON
- Add CI pipelines for aarch64 using GCC cross compiler and QEMU user mode emulation

### CODE REVIEW

**Review by guillaumekln — COMMENTED** [review]
Thanks, the integration looks very good!

In addition to the few comments below, I have some general questions:

* In the related issue I mentioned the project [ARM Compute Library](https://github.com/ARM-software/ComputeLibrary). Do you have any experience with it? If yes, how does it compare with Apple Accelerate and OpenBLAS?
* Do you plan to add 8-bit support in the future?

File: README.md [hunk 1]
```diff
@@ -276,6 +278,8 @@ Some build options require external dependencies:
   * [Intel MKL](https://software.intel.com/en-us/mkl) (>=2019.5)
 * `-DWITH_DNNL=ON` requires:
   * [oneDNN](https://github.com/oneapi-src/oneDNN) (>=1.5)
+* `-DWITH_OPENBLAS=ON` requires:
+  * [OpenBLAS](https://github.com/xianyi/OpenBLAS)
```
[the remark this answers is not in the document]
Comment (keichi) [reply README.md:284]: Sure I've added an entry for Accelerate.

File: CMakeLists.txt [hunk 2]
```diff
@@ -134,21 +136,27 @@ macro(ct2_compile_kernels_for_isa isa flag)
     src/cpu/kernels.cc
     ${CMAKE_CURRENT_BINARY_DIR}/kernels_${isa}.cc
     COPYONLY)
-  set_source_files_properties(
-    ${CMAKE_CURRENT_BINARY_DIR}/kernels_${isa}.cc
-    PROPERTIES COMPILE_FLAGS ${flag})
+  if(NOT flag STREQUAL "")
```
[the remark this answers is not in the document]
Comment (keichi) [reply CMakeLists.txt]: Ooops, that was left out when I was experimenting around with the CMake script.

**Comment by keichi** [comment]
> * In the related issue I mentioned the project [ARM Compute Library](https://github.com/ARM-software/ComputeLibrary). Do you have any experience with it? If yes, how does it compare with Apple Accelerate and OpenBLAS?

Unfortunately, I don't have any experience with the ARM Compute Library. It seems to be a higher-level library than BLAS  (it has its own tensor type, memory manager, etc.) so it would require more work than just replacing calls to `cblas_gemm()`. I could look into it in the future.

I do believe Apple's Accelerate is the fastest among them on Apple's hardware since it utilizes the matrix multiplication coprocessor (see https://gist.github.com/dougallj/7a75a3be1ec69ca550e7c36dc75e0d6f).

> * Do you plan to add 8-bit support in the future?

INT8 should already be supported by DNNL. If you see the logs for `build-and-test-cpp-aarch64`:

```
[ct2_verbose] CPU: ARM (SSE4.1=false, AVX=false, AVX2=false, NEON=true)
[ct2_verbose]  - Selected ISA: NEON
[ct2_verbose]  - Use Intel MKL: false
[ct2_verbose]  - SGEMM backend: DNNL
[ct2_verbose]  - GEMM_S16 backend: NONE
[ct2_verbose]  - GEMM_S8 backend: DNNL (u8s8 preferred: true)
[ct2_verbose]  - Use packed GEMM: false
```

Both Accelerate and OpenBLAS do not support INT8, FLOAT16, etc.

**Comment by guillaumekln** [comment]
Thanks for the info and the PR!

**Comment by WilliamVJacob** [comment]
Hi, I wish to cross compile ctranslate2 for ARM embedded board. Can you please explain the how do I do it.

Also what are the dependency for ctranslate2 for ARM

**Comment by keichi** [comment]
You either need OpenBLAS or oneDNN on an Linux aarch64 system.

The steps needed for cross compilation would depend on your system and toolchain. This excerpt from our CI script that cross compiles CTranslate2 with OpenBLAS might help.

https://github.com/OpenNMT/CTranslate2/blob/e11870feeb56eec1c1ca4ab83394f50b4402d1bd/.github/workflows/ci.yml#L123-L142

**Comment by neso613** [comment]
Can I compile for R-Pi 32 bit also.
Support needed.

**Comment by keichi** [comment]
32-bit ARM is currently unsupported and untested, but you might be able to compile for. Since oneDNN does not support 32-bit ARM, you should use OpenBLAS as the GEMM backend. Another option is to install 64-bit Raspberry Pi OS or Ubuntu if you are using recent RasPis.

**Comment by neso613** [comment]
> 32-bit ARM is currently unsupported and untested, but you might be able to compile for. Since oneDNN does not support 32-bit ARM, you should use OpenBLAS as the GEMM backend. Another option is to install 64-bit Raspberry Pi OS or Ubuntu if you are using recent RasPis.
I am stuck with this error:
CTranslate2/src/cpu/primitives.cc: In static member function ‘static T ctranslate2::primitives::sum(const T*, ctranslate2::dim_t) [with T = T; ctranslate2::Device D = (ctranslate2::Device)0]’:
/home/pi/neha/CTranslate2/src/cpu/primitives.cc:81:45: error: ‘ISA’ was not declared in this scope
CPU_ISA_DISPATCH((sum = cpu::reduce_sum(array, size)));
^~~
https://www.gitmemory.com/issue/OpenNMT/CTranslate2/370/756747531

**Comment by keichi** [comment]
Can you add `-DENABLE_CPU_DISPATCH=OFF` when invoking CMake?

**Comment by neso613** [comment]
> -DENABLE_CPU_DISPATCH=OFF

yes, IS_CPU_DISPATCH issue is solved.
But,
  ^~~~~~~~~~~~~~~~~
[ 86%] Building CXX object CMakeFiles/ctranslate2.dir/src/translator.cc.o
[ 87%] Building CXX object CMakeFiles/ctranslate2.dir/src/translator_pool.cc.o
[ 89%] Building CXX object CMakeFiles/ctranslate2.dir/src/types.cc.o
[ 91%] Building CXX object CMakeFiles/ctranslate2.dir/src/utils.cc.o
[ 93%] Building CXX object CMakeFiles/ctranslate2.dir/src/vocabulary.cc.o
[ 94%] Building CXX object CMakeFiles/ctranslate2.dir/src/vocabulary_map.cc.o
[ 96%] Linking CXX shared library libctranslate2.so
[ 96%] Built target ctranslate2
Scanning dependencies of target translate
[ 98%] Building CXX object cli/CMakeFiles/translate.dir/translate.cc.o
[100%] Linking CXX executable translate
/usr/bin/ld: ../libctranslate2.so: undefined reference to `ctranslate2::cpu::cpu_supports_avx2()'
/usr/bin/ld: ../libctranslate2.so: undefined reference to `ctranslate2::cpu::cpu_supports_avx()'
/usr/bin/ld: ../libctranslate2.so: undefined reference to `ctranslate2::cpu::cpu_supports_neon()'
/usr/bin/ld: ../libctranslate2.so: undefined reference to `ctranslate2::cpu::cpu_vendor[abi:cxx11]()'
/usr/bin/ld: ../libctranslate2.so: undefined reference to `ctranslate2::cpu::cpu_supports_sse41()'
collect2: error: ld returned 1 exit status
make[2]: *** [cli/CMakeFiles/translate.dir/build.make:86: cli/translate] Error 1
make[1]: *** [CMakeFiles/Makefile2:123: cli/CMakeFiles/translate.dir/all] Error 2
make: *** [Makefile:130: all] Error 2

**Comment by keichi** [comment]
You can replace the following line with `#elif 1` to pretend you're on ARM64 but it's pretty hacky... 
https://github.com/OpenNMT/CTranslate2/blob/974f8db56df3f0e1a59536a42afa93632d141108/src/cpu/cpu_info.cc#L105

**Comment by neso613** [comment]
Can I compile Ctranslate2 for GPU on arm?

**Comment by keichi** [comment]
Do you have an NVIDIA GPU that supports CUDA? Otherwise it’s not possible.

**Comment by neso613** [comment]
> Do you have an NVIDIA GPU that supports CUDA? Otherwise it’s not possible.

Yes, I have.
For this, I have tried to get intel MKL. https://github.com/OpenNMT/CTranslate2#install-intel-mkl-optional-for-gpu-only-builds
sudo sh -c 'echo "deb https://apt.repos.intel.com/oneapi all main" > /etc/apt/sources.list.d/oneAPI.list'
Then , I am getting this - 
`N: Skipping acquire of configured file 'main/binary-arm64/Packages' as repository 'https://apt.repos.intel.com/oneapi all InRelease' doesn't support architecture 'arm64'`

**Comment by keichi** [comment]
MKL is optional for GPU builds as mentioned in the README. Also, MKL does not run on ARM. It's a library designed for Intel CPUs.

You should disable the MKL backend and enable the CUDA backend instead when configuring CTranslate2: `-DWITH_MKL=OFF -DWITH_CUDA=ON`.

**Comment by neso613** [comment]
> MKL is optional for GPU builds as mentioned in the README. Also, MKL does not run on ARM. It's a library designed for Intel CPUs.
> 
> You should disable the MKL backend and enable the CUDA backend instead when configuring CTranslate2: `-DWITH_MKL=OFF -DWITH_CUDA=ON`.

**Comment by neso613** [comment]
> MKL is optional for GPU builds as mentioned in the README. Also, MKL does not run on ARM. It's a library designed for Intel CPUs.
> 
> You should disable the MKL backend and enable the CUDA backend instead when configuring CTranslate2: `-DWITH_MKL=OFF -DWITH_CUDA=ON`.

tried --
-- Compiling for multiple CPU ISA and enabling runtime dispatch
-- A library with BLAS API found.
-- Autodetected CUDA architecture(s): 7.2
-- NVCC compilation flags: -std=c++11;-gencode;arch=compute_72,code=sm_72
-- Found CUB include directory: /home/gpu-ctranslate/CTranslate2/third_party/cub
-- Found Thrust include directory: /home/gpu-ctranslate/CTranslate2/third_party/thrust
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
CUDA_cublas_device_LIBRARY (ADVANCED)
linked by target "ctranslate2" in directory /home/gpu-ctranslate/CTranslate2

-- Configuring incomplete, errors occurred!
See also "/home/gpu-ctranslate/CTranslate2/CMakeFiles/CMakeOutput.log".
See also "/home/gpu-ctranslate/CTranslate2/CMakeFiles/CMakeError.log".

**Comment by keichi** [comment]
@neso613 Can you open a new issue? Your problem seems unrelated to ARM.

**Comment by neso613** [comment]
> @neso613 Can you open a new issue? Your problem seems unrelated to ARM.

ok. I will but this issue is on Nvidia jetson board.

**Comment by yywing** [comment]
`pip install ctranslate2` can not install.

`Could not find a version that satisfies the requirement ctranslate2`

**Comment by reisub0** [comment]
Forgive me for resurrecting this old thread, but how do I confirm that I'm indeed running on Apple Accelerate (I'm using an M1 device)

If I install `ctranslate2-3.8.0-cp311-cp311-macosx_11_0_arm64.whl` from `pip` and I try to run Whisper for eg, like so:

```
model = WhisperModel(model_path, device="cpu", compute_type="auto")
```

Is it going to be guaranteed to use the Accelerate backend?

**Comment by guillaumekln** [comment]
Yes, Accelerate is the only backend available in macOS ARM binaries so it is always used.

You can enable the info logs before loading the model to find more information about the selected backend.

```python
import logging
import ctranslate2
ctranslate2.set_log_level(logging.INFO)

model = WhisperModel(model_path, device="cpu", compute_type="auto")
```

**Comment by zara0m** [comment]
> Yes, Accelerate is the only backend available in macOS ARM binaries so it is always used.
> 
> You can enable the info logs before loading the model to find more information about the selected backend.
> 
> ```python
> import logging
> import ctranslate2
> ctranslate2.set_log_level(logging.INFO)
> 
> model = WhisperModel(model_path, device="cpu", compute_type="auto")
> ```

Hello,

I have tried this, with cpu_thread=8, and my log looks like this:

[ctranslate2] [thread 124186] [warning] The number of threads (intra_threads) is ignored in this build
[ctranslate2] [thread 124186] [info] CPU: ARM (NEON=true)
[ctranslate2] [thread 124186] [info]  - Selected ISA: NEON
[ctranslate2] [thread 124186] [info]  - Use Intel MKL: false
[ctranslate2] [thread 124186] [info]  - SGEMM backend: Accelerate (packed: false)
[ctranslate2] [thread 124186] [info]  - GEMM_S16 backend: none (packed: false)
[ctranslate2] [thread 124186] [info]  - GEMM_S8 backend: none (packed: false, u8s8 preferred: false)
[ctranslate2] [thread 124186] [info] Loaded model small-int8[/](https://file+.vscode-resource.vscode-cdn.net/) on device cpu:0
[ctranslate2] [thread 124186] [info]  - Binary version: 6
[ctranslate2] [thread 124186] [info]  - Model specification revision: 3
[ctranslate2] [thread 124186] [info]  - Selected compute type: float32


So I think it uses accelerate, and the cpu usage is 120%. I also tried installing ctranslate2 by building from source using -DWITH_ACCELERATE=ON, and the log looks same.

I wonder if there is anything that I can do to improve performance on Mac M1 Max with 10 cores?

I appreciate it if you could help me.

**Comment by guillaumekln** [comment]
Hi,

See this issue https://github.com/OpenNMT/CTranslate2/issues/1137.

Apple Accelerate does not use multiple threads so this option has no effect. One solution is to compile with `-DWITH_RUY` and run the model with the "int8" compute type.

**Comment by zara0m** [comment]
Thank you for your help.

**Comment by ururk** [comment]
> Apple Accelerate does not use multiple threads so this option has no effect. One solution is to compile with `-DWITH_RUY=ON` and run the model with the "int8" compute type.

Will the prebuilt binaries be compiled with `DWITH_RUY=ON`? Or do we need to build this ourselves?

**Comment by guillaumekln** [comment]
The Ruy backend will be included in the next version of the prebuilt Python wheels for macOS ARM: https://github.com/OpenNMT/CTranslate2/commit/773cea82e003e90972121a458c0369f3e2cf5df7

**Pull request merged**